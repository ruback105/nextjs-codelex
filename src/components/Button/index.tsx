import React, { ButtonHTMLAttributes, ReactNode } from "react";
import classNames from "classnames";

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  icon?: ReactNode;
  children: ReactNode;
  primary?: boolean;
  hover?: boolean;
};

export const buttonClassName = (primary?: boolean, hover?: boolean) =>
  classNames("whitespace-nowrap p-4 rounded-full min-w-[200px] font-semibold", {
    "bg-[#8692A6] text-white": primary,
    "hover:bg-[#8692A6] hover:text-white": hover,
  });

const Button: React.FC<ButtonProps> = ({
  icon,
  children,
  primary,
  hover,
  ...rest
}) => {
  return (
    <button
      className={classNames(
        buttonClassName(primary, hover),
        "gap-x-3 items-center flex justify-center"
      )}
      {...rest}
      data-testid="button"
    >
      {icon}
      {children}
    </button>
  );
};

export default Button;
