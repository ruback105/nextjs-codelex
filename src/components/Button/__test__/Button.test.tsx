import { fireEvent, render } from "@testing-library/react";
import Button from "..";

describe("Button", () => {
  const textChildren = "Click";
  const jsxChildren = <p data-testid="jsx-children">{textChildren}</p>;

  it("render without crashing", () => {
    render(<Button>{textChildren}</Button>);
  });

  it("render text children", () => {
    const { getByText } = render(<Button>{textChildren}</Button>);

    expect(getByText(textChildren)).toBeInTheDocument();
  });

  it("render jsx children", () => {
    const { getByTestId } = render(<Button>{jsxChildren}</Button>);

    expect(getByTestId("jsx-children").innerHTML).toEqual(textChildren);
  });

  it("render default styles if no props are passed", () => {
    const { getByTestId } = render(<Button>{textChildren}</Button>);

    expect(getByTestId("button").className).toEqual(
      "whitespace-nowrap p-4 rounded-full min-w-[200px] font-semibold gap-x-3 items-center flex justify-center"
    );
  });

  it("render primary styles if primary flag is passed", () => {
    const { getByTestId } = render(<Button primary>{textChildren}</Button>);

    expect(getByTestId("button").className).toEqual(
      "whitespace-nowrap p-4 rounded-full min-w-[200px] font-semibold bg-[#8692A6] text-white gap-x-3 items-center flex justify-center"
    );
  });

  it("triggers onClick when clicked", () => {
    const onClick = jest.fn();

    const { getByTestId } = render(
      <Button onClick={onClick}>{textChildren}</Button>
    );

    fireEvent.click(getByTestId("button"));

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("default Button snapshot", () => {
    const { container } = render(<Button>{textChildren}</Button>);

    expect(container).toMatchSnapshot();
  });
});
