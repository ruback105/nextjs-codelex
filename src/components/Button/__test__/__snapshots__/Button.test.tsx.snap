// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default Button snapshot 1`] = `
<div>
  <button
    class="whitespace-nowrap p-4 rounded-full min-w-[200px] font-semibold gap-x-3 items-center flex justify-center"
    data-testid="button"
  >
    Click
  </button>
</div>
`;
